### ISSUE TITLE
Refresh view after decrypting

### ISSUE THREAD

**Issue by Chris00** [djcb/mu#876]
When I first decrypt a message, it displays it as attachments (see first image).  However, if I move to another message and then come back to the encrypted one, it displays as it should (second image).  It would be nice if the refresh was done automatically.

![decrypt1](https://cloud.githubusercontent.com/assets/1255665/16610938/b3eb48a2-435e-11e6-91a2-134620395953.png)
![decrypt2](https://cloud.githubusercontent.com/assets/1255665/16610978/fe1864b4-435e-11e6-95c0-c39be1d7187b.png)

**Comment by Chris00** [comment]
The same thing happens for attached images.  When one displays the message for the first time, the images are only shown as attachments.  Subsequent displays (for example passing to another message and coming back), show the images inlined (what I expect as I have `(setq mu4e-view-show-images t)`).  Should I open a new issue for this?

**Comment by toothbrush** [comment]
I also have/had this problem.  I'm not 100% sure that it's solved, but it appears to be (i don't receive that much crypted mail, so haven't had the opportunity to test extensively) fixed after adding the following into my config:

```elisp
  (add-hook 'mu4e-view-mode-hook 'epa-mail-mode)
```

Let me know if that helps or if it's just a red herring.

**Comment by djcb** [comment]
There were some recent changes with crypto-support -- most importantly, preferring `gpg2` over `gpg`. In any case, I am unable to reproduce the mentioned. 

As for images, there's #447 for that.

**Comment by Chris00** [comment]
@toothbrush It does not work.

**Comment by Chris00** [comment]
@djcb The issue is triggered simply by sending an encrypted message to myself.  Maybe Emacs does not wait that the message is decrypted before displaying it?

**Comment by toothbrush** [comment]
Agreed, for me the issue also still exists.

On 2016-08-03 at 18:18 AEST+1000, quoth Christophe Troestler:

**Comment by djcb** [comment]
[ meta-comment: when filling issue, please use the template, since we're missing essential information here about versions etc.]

Did you try the latest mu (git)? What version of gpg are you using? Does decrypting use the command-line mu work?

**Comment by Chris00** [comment]
Sorry for forgetting.  I use a Git clone of `mu`.  `gpg --version` says `1.4.20` (I have to keep it as this version is used by essential tools on the system) but I also have `gpg2`:
```
$ gpg2 --version
gpg (GnuPG) 2.1.11
libgcrypt 1.7.2-beta
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by Chris00** [comment]
```
$ mu view --decrypt ~/Maildir/UMONS/inbox/cur/1470260532.9837_1.poincare,U=15251:2,S
From: Christophe Troestler <Christophe.Troestler@umons.ac.be>
To: Christophe.Troestler@umons.ac.be
Subject: test
Date: Wed 03 Aug 2016 22:41:56 BST
mu: error decrypting part: verification failed
mu: error decrypting part: verification failed


Encrypt
```
Not sure how to know which version of gpg was used.

**Comment by Chris00** [comment]
(The above message I was trying to decrypt was composed with mu4e.)

**Comment by toothbrush** [comment]
Perchance i discovered something that may be useful in solving the issue.  If i mark a message as read before trying to open it in mu4e (e.g. in a mail agent that doesn't understand GPG) then mu4e manages to open and decrypt it first time around.  I suspect that the "rename Maildir file to append `S`" is somehow preventing the decryption (is that `epa`'s responsibility?) from taking place.

**Comment by Chris00** [comment]
The same mechanism may be the cause of bugs https://github.com/djcb/mu/issues/447 and https://github.com/djcb/mu/issues/474

**Comment by spacecowboy** [comment]
I did some digging on this tonight and think I found the problem. When flags are updated, `mu4e-view` is called a second time. But this time without any decryption (or show images) flags. Here's a diff with an experimental change:

```diff
--- a/mu4e/mu4e-headers.el
+++ b/mu4e/mu4e-headers.el
@@ -353,7 +353,8 @@ headers."
          ;; update it; that way, the flags can be updated, as well as the path
          ;; (which is useful for viewing the raw message)
          (when (mu4e~headers-view-this-message-p docid)
-           (mu4e-view msg mu4e~headers-buffer))
+           ; t: always decrypt for this test
+           (mu4e~proc-view docid mu4e-view-show-images t))
          ;; now, if this update was about *moving* a message, we don't show it
          ;; anymore (of course, we cannot be sure if the message really no
          ;; longer matches the query, but this seem a good heuristic.  if it
```

This change makes sure that mails show images and decrypt messages upon refresh. I've tested both.

It's not a good fix though since this means the message is decrypted twice and since I'm using a Yubikey (smartcard), decrypting takes a long time (1 second or so). And doing it twice just seems wrong anyway. Must be a better way of first updating flags and then opening the message.

I could digg further and come up with a realistic fix unless you wanna take it from here @djcb

**Comment by semente** [comment]
@spacecowboy's patch works fine! thanks!

**Comment by yurkobb** [comment]
Experiencing the same issue on commit 82743b31.

**Comment by djcb** [comment]
Okay, this is fixed with the new (experimental) view mode -- `(setq mu4e-view-use-gnus t)` (before starting mu4e). Finally!

**Comment by spacecowboy** [comment]
Finally fixed with cd649efb6be6cfc1a3748781467c30996b5c20a9 !
